Section#within - solution for #21

tgaff opened the pull request:
## Overview

Hi Luke,
This is one of the potential ways to resolve #21 that I discussed so long ago.  I'm sorry it took this long to get you anything for this.  Even more so, since this turned out to be much simpler than I anticipated.

#### Issue Review #21 
In #21 we had the problem that the following was OK:

```rb
page.my_section do |sec|
  expect(sec).to have_secs_element
end
```

But the following silently ignored the block because when the block is executed in the above, it's executed as part of the `initialize` method.
```rb
page.my_sections.first do |sec|
  expect(sec).to have_secs_element
end
```

**In this case the sections are initialized and put into an Array.  Initialization has passed and Array#first doesn't use a block.**

#### This change
With this change you can do basically the same thing by calling `#within`.
```rb
page.my_sections.first.within do |sec|
  expect(sec).to have_secs_element
end
```

### Other Alternatives

We could implement a `SectionsCollection` class that serves as the `Array` for Sections.  
We would need the basic methods such as:
* `first`
* `last`
* `map`
* `each`

These could take a block as needed.

Capybara does something similar with [Capybara::Result](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Result) which is the result class for `page.all('.locator')`

That would obviously be more complex and I'm not sure we have a strong need for it yet, thoughts? Regardless, I think this actually would be useful as a start for that, as we'd still want a good way to execute easily in the context of a section at will.

## TODO / Questions

* docs -   Figured I would see how you felt about it first @luke-hill 
* do we want to discourage the use of sending a block to Section#new (which is maybe not obvious to many users) and instead encourage using #within ?
* I want to give this a test on one of my repos JIC
* should I bump the version?
* I added a pry **dev** dependency - it's a very common gem and I end up pulling it in every time I work on site-prism then undoing that add.

luke-hill commented on lib/site_prism/section.rb:
```diff
@@ -31,6 +31,10 @@ def initialize(parent, root_element)
       Capybara.within(@root_element) { yield(self) } if block_given?
```
lets refactor this line

luke-hill replied:
ping here.

luke-hill commented:
p.s. ignore the failure. I need to do a slight bit of refactoring.

Since the latest 2.5 cut came out, the runtimes have actually **improved**. I just need to make a couple of refactors to the test codebase and isolate the runtimes, because it's loads of duplication I can remove.

luke-hill commented:
Hi @tgaff 

What do you want to do with this, I've only put a small refactor requested, but this is essentially RtM unless you want to make some additional fixes?

tgaff commented:
I know I'm forever slow.  Sorry.  I'll write some docs and make the change you requested!  
I wanted to test this in a bigger code-base but haven't been able to yet (need to do some upgrades on that).

luke-hill commented:
Ping @tgaff

tgaff commented:
Sorry Luke, changed jobs and don't have that bigger code-base to test on now.  

I'll write up the docs and get this in.  Sorry once again.

tgaff commented:
Ok, I added a little documentation and rebased it.  Tests are still failing but I think its the same failure as on master.  Maybe a Capybara API update?

luke-hill commented:
@tgaff I'm busy for 1 or 2 weeks. For now can you add an upper restriction of say 3.29 to capybara and see if that helps.

It's one test and one missing method, so that's likely it.

tgaff commented:
Luke your thoughts on 3.30 being the issue were correct.  Thanks!

tgaff commented on README.md:
````diff
@@ -937,6 +936,16 @@ Then(/^the home page menu contains a link to the various search functions$/) do
 end
 ```
 
+Note that on individual sections it's possible to pass a block directly to the section without using `within`.  Because the block is executed only during Section initialization this won't work when accessing a single Section from an array of Sections.  For that reason we recommend using `within` which works in either case.
````
✔️

luke-hill commented:
Cheers travis. I think I'll dig into this issue a bit more and maybe put a more strict check on gem version, but this PR is rtm!
